File: generate_assets.js
import Jimp from 'jimp';
import path from 'path';
import fs from 'fs';

const logoPath = './frontend/public/images/logo.png';
const publicDir = './frontend/public';

async function main() {
  if (!fs.existsSync(logoPath)) {
    console.error(`Logo not found at ${logoPath}`);
    return;
  }

  console.log(`Found logo at ${logoPath}. Generating assets...`);
  const image = await Jimp.read(logoPath);
  
  // favicon-16x16
  const fav16 = image.clone().resize(16, 16);
  await fav16.write(path.join(publicDir, 'favicon-16x16.png'));
  console.log('Generated favicon-16x16.png');

  // favicon-32x32
  const fav32 = image.clone().resize(32, 32);
  await fav32.write(path.join(publicDir, 'favicon-32x32.png'));
  console.log('Generated favicon-32x32.png');

  // favicon-48x48 (Google Search Preferred multiple of 48)
  const fav48 = image.clone().resize(48, 48);
  await fav48.write(path.join(publicDir, 'favicon-48x48.png'));
  console.log('Generated favicon-48x48.png');

  // apple-touch-icon (180x180)
  const apple = image.clone().resize(180, 180);
  await apple.write(path.join(publicDir, 'apple-touch-icon.png'));
  console.log('Generated apple-touch-icon.png');

  // android-chrome-192 (192x192)
  const chrome192 = image.clone().resize(192, 192);
  await chrome192.write(path.join(publicDir, 'android-chrome-192x192.png'));
  console.log('Generated android-chrome-192x192.png');

  // android-chrome-512 (512x512)
  const chrome512 = image.clone().resize(512, 512);
  await chrome512.write(path.join(publicDir, 'android-chrome-512x512.png'));
  console.log('Generated android-chrome-512x512.png');

  // For favicon.ico, we output 32x32
  const ico = image.clone().resize(32, 32);
  await ico.write(path.join(publicDir, 'favicon.ico'));
  console.log('Generated favicon.ico');

  // og-image (1200x630) with parchment (#E9DFC9) background
  // #E9DFC9 in hex is 0xE9DFC9FF in Jimp color representation
  const ogBg = await new Jimp(1200, 630, 0xE9DFC9FF);
  
  // Resize logo to fit inside 1200x630
  const logoResized = image.clone();
  logoResized.resize(Jimp.AUTO, 350);
  if (logoResized.getWidth() > 800) {
    logoResized.resize(800, Jimp.AUTO);
  }

  const x = Math.floor((1200 - logoResized.getWidth()) / 2);
  const y = Math.floor((630 - logoResized.getHeight()) / 2);
  
  ogBg.composite(logoResized, x, y);
  await ogBg.write(path.join(publicDir, 'og-image.jpg'));
  console.log('Generated og-image.jpg');

  console.log('All favicon and social preview assets generated successfully!');
}

main().catch(err => {
  console.error(err);
});

File: generate_assets.py
import os
import sys

def main():
    try:
        from PIL import Image, ImageOps
    except ImportError:
        print("Pillow is not installed. Installing Pillow...")
        import subprocess
        subprocess.check_call([sys.executable, "-m", "pip", "install", "Pillow"])
        from PIL import Image, ImageOps

    logo_path = os.path.join("frontend", "public", "images", "logo.png")
    public_dir = os.path.join("frontend", "public")

    if not os.path.exists(logo_path):
        print(f"Error: logo.png not found at {logo_path}")
        return

    print(f"Found logo.png at {logo_path}. Generating assets...")

    # Load logo
    img = Image.open(logo_path)

    # 1. favicon-16x16.png
    fav16 = img.resize((16, 16), Image.Resampling.LANCZOS)
    fav16.save(os.path.join(public_dir, "favicon-16x16.png"))
    print("Generated favicon-16x16.png")

    # 2. favicon-32x32.png
    fav32 = img.resize((32, 32), Image.Resampling.LANCZOS)
    fav32.save(os.path.join(public_dir, "favicon-32x32.png"))
    print("Generated favicon-32x32.png")

    # 3. apple-touch-icon.png (180x180)
    apple_icon = img.resize((180, 180), Image.Resampling.LANCZOS)
    apple_icon.save(os.path.join(public_dir, "apple-touch-icon.png"))
    print("Generated apple-touch-icon.png")

    # 4. favicon.ico (multi-resolution / contains 16x16 and 32x32)
    img.save(os.path.join(public_dir, "favicon.ico"), format="ICO", sizes=[(16, 16), (32, 32)])
    print("Generated favicon.ico")

    # 5. og-image.jpg (1200x630)
    # Background color: parchment (#E9DFC9) -> (233, 223, 201)
    bg_color = (233, 223, 201)
    og_img = Image.new("RGB", (1200, 630), bg_color)

    # Resize logo to fit nicely inside 1200x630 (e.g. height of 400px)
    logo_w, logo_h = img.size
    target_h = 350
    target_w = int(logo_w * (target_h / logo_h))
    
    # Ensure it's not too wide
    if target_w > 800:
        target_w = 800
        target_h = int(logo_h * (target_w / logo_w))

    resized_logo = img.resize((target_w, target_h), Image.Resampling.LANCZOS)

    # Paste centered
    x = (1200 - target_w) // 2
    y = (630 - target_h) // 2

    # If logo has alpha channel, use it as mask
    if resized_logo.mode in ('RGBA', 'LA') or (resized_logo.mode == 'P' and 'transparency' in resized_logo.info):
        # Convert to RGBA if not already
        resized_logo = resized_logo.convert('RGBA')
        og_img.paste(resized_logo, (x, y), resized_logo)
    else:
        og_img.paste(resized_logo, (x, y))

    og_img.save(os.path.join(public_dir, "og-image.jpg"), "JPEG", quality=90)
    print("Generated og-image.jpg")

    print("All assets generated successfully!")

if __name__ == "__main__":
    main()
